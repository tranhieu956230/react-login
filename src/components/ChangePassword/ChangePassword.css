.change-password__title {
  font-size: 3rem;
  font-weight: 500;
  text-align: center;
  margin-top: 2rem;
  color: rgba(41, 41, 41, 0.82);
  margin-bottom: 2rem;
}

.change-password__form {
  max-width: 55rem;
  display: grid;
  justify-items: center;
  width: 60rem;
}

.change-password__container {
  flex: 1;
  background-color: white;
  display: flex;
  justify-content: center;
  align-items: center;
}
