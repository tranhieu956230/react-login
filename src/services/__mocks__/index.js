export const activateAccount = jest.fn();
