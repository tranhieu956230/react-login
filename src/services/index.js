const delay = milliseconds => {
  return new Promise(resolve => {
    setTimeout(() => resolve(), milliseconds);
  });
};

let userInfo = {
  username: "",
  password: "",
  name: ""
};

export const login = async (username, password) => {
  await delay(200);
  if (username === userInfo.username && password === userInfo.password) {
    return {
      success: true,
      access_token: "ACCESS_TOKEN",
      code: 200
    };
  }
  return {
    success: false,
    message: "Invalid username or password.",
    code: 400
  };
};

export const signUp = async (email, password, confirmPassword, name) => {
  await delay(200);
  if (password === confirmPassword) {
    userInfo.username = email;
    userInfo.password = password;
    return {
      success: true,
      code: 200
    };
  }
  return {
    success: false,
    message: "Password does not match."
  };
};
