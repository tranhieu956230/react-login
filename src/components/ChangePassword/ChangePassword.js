import React, { useState } from "react";
import styled from "styled-components";
import { setGlobal, useGlobal } from "reactn";

import Header from "components/shared/Header";
import { changePassword } from "services";
import InputValidate from "components/shared/InputValidate";
import validateUtil from "utils/validate.js";
import PrimaryButton from "components/shared/PrimaryButton";

const ChangePassword = props => {
  const [newPassword, setNewPassword] = useState("");
  const [newPasswordConfirm, setNewPasswordConfirm] = useState("");
  const [oldPassword, setOldPassword] = useState("");
  const [isValidOldPassword, setIsValidOldPassword] = useState(true);
  const [isValidNewPassword, setIsValidNewPassword] = useState(true);
  const [isValidNewPasswordConfirm, setIsValidNewPasswordConfirm] = useState(
    true
  );
  const [isChangePasswordRequest, setIsChangePasswordRequest] = useState(false);
  const [accessToken] = useGlobal("accessToken");

  const handleSubmit = e => {
    e.preventDefault();
    let valid = true;

    if (!newPasswordConfirm || !isValidNewPasswordConfirm) {
      valid = false;
      setIsValidNewPasswordConfirm(false);
    }

    if (!newPassword || !isValidNewPassword) {
      valid = false;
      setIsValidNewPassword(false);
    }

    if (!oldPassword || !isValidOldPassword) {
      valid = false;
      setIsValidOldPassword(false);
    }

    if (valid && !isChangePasswordRequest) {
      let t = setTimeout(() => {
        setIsChangePasswordRequest(true);
      }, 200);

      changePassword(
        oldPassword,
        newPassword,
        newPasswordConfirm,
        accessToken
      ).then(response => {
        clearTimeout(t);
        setIsChangePasswordRequest(false);
        if (response.code in ["SUCCESS", "UNAUTHORIZED"]) {
          setGlobal({
            isLoggedIn: false,
            access_token: ""
          });
          localStorage.removeItem("access_token");
          props.history.push("/");
        }
      });
    }
  };

  const handleOldPasswordChange = e => {
    let oldPassword = e.target.value;
    setOldPassword(oldPassword);
    setIsValidOldPassword(validateUtil.isValidPassword(oldPassword));
  };

  const handleNewPasswordChange = e => {
    let newPassword = e.target.value;
    setNewPassword(newPassword);
    setIsValidNewPassword(validateUtil.isValidPassword(newPassword));
    setIsValidNewPasswordConfirm(newPassword === newPasswordConfirm);
  };

  const handleNewPasswordConfirmChange = e => {
    setNewPasswordConfirm(e.target.value);
    setIsValidNewPasswordConfirm(newPassword === e.target.value);
  };

  return (
    <div>
      <Header history={props.history} />
      <Text>Change Password</Text>
      <Main>
        <FormWrapper onSubmit={handleSubmit}>
          <InputValidate
            placeholder={"Old Password"}
            type={"password"}
            onChange={handleOldPasswordChange}
            value={oldPassword}
            errorMessage={
              "Password must has 1 lowercase, 1 uppercase letter, 1 number and at least 8 characters in length"
            }
            valid={isValidOldPassword}
          />
          <InputValidate
            placeholder={"New Password"}
            type={"password"}
            onChange={handleNewPasswordChange}
            value={newPassword}
            errorMessage={
              "Password must has 1 lowercase, 1 uppercase letter, 1 number and at least 8 characters in length"
            }
            valid={isValidNewPassword}
          />
          <InputValidate
            placeholder={"Confirm new password"}
            type={"password"}
            onChange={handleNewPasswordConfirmChange}
            value={newPasswordConfirm}
            errorMessage={"Password not match"}
            valid={isValidNewPasswordConfirm}
          />
          <PrimaryButton isLoading={isChangePasswordRequest} text={"Submit"} />
        </FormWrapper>
      </Main>
    </div>
  );
};

const Text = styled.h1`
  font-size: 3rem;
  font-weight: 500;
  text-align: center;
  margin-top: 2rem;
  color: rgba(41, 41, 41, 0.82);
  margin-bottom: 2rem;
`;

const FormWrapper = styled.form`
  max-width: 55rem;
  display: grid;
  justify-items: center;
  width: 60rem;
`;

const Main = styled.div`
  flex: 1;
  background-color: white;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export default ChangePassword;
