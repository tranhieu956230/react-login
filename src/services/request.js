import { activateAccount } from "services";

export default activateAccount;
