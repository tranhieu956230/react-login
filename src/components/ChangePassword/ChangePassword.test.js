import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import { BrowserRouter, Route } from "react-router-dom";

import ChangePassword from "./";
import { changePassword } from "services";

jest.mock("services", () => ({
  changePassword: jest.fn()
}));

describe("Test ChangePassword Shallow", () => {
  it("Should render with no error", () => {
    const wrapper = shallow(<ChangePassword />);
    expect(wrapper).toBeDefined();
  });

  it("Should match snapshot", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <Route render={props => <ChangePassword {...props} />} />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Test ChangePassword Mount", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <BrowserRouter>
        <Route render={props => <ChangePassword {...props} />} />
      </BrowserRouter>
    );
  });

  afterEach(() => {
    changePassword.mockReset();
    wrapper.unmount();
  });

  it("Should not call api", async done => {
    wrapper.find("PrimaryButton").simulate("click");
    setTimeout(() => {
      expect(changePassword).toHaveBeenCalledTimes(0);
      done();
    });
  });

  it("Should call api with arguments", async done => {
    const oldPassword = wrapper.find(".input__field").at(0);
    const newPassword = wrapper.find(".input__field").at(1);
    const passwordConfirm = wrapper.find(".input__field").at(2);

    const oldPasswordValue = "oldPassword11235";
    const newPasswordValue = "newPassword11235";
    const passwordConfirmValue = "newPassword11235";

    oldPassword.simulate("change", { target: { value: oldPasswordValue } });
    newPassword.simulate("change", { target: { value: newPasswordValue } });
    passwordConfirm.simulate("change", { target: { value: passwordConfirmValue } });

    changePassword.mockImplementationOnce((a, b, c) => {
      return new Promise(resolve => {
        resolve({
          code: "SUCCESS"
        });
      });
    });

    wrapper.find(".change-password__form").simulate("submit");
    setTimeout(() => {
      expect(changePassword).toBeCalledWith(oldPasswordValue, newPasswordValue, undefined);
      done();
    });
  });
});
