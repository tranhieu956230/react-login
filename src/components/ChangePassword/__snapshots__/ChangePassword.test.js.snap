// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ChangePassword Shallow Should match snapshot 1`] = `
<div>
  <div
    className="header__container"
  >
    <button
      className="header__button"
      onClick={[Function]}
    >
      Settings
    </button>
    <ul
      className="header__drop-down "
    >
      <li
        className="header__drop-down-item"
      >
        <a
          className="header__link"
          href="/change-password"
          onClick={[Function]}
        >
          Change your password
        </a>
      </li>
      <li
        className="header__drop-down-item"
        onClick={[Function]}
      >
        Logout
      </li>
    </ul>
  </div>
  <h1
    className="change-password__title"
  >
    Change Password
  </h1>
  <div
    className="change-password__container"
  >
    <form
      className="change-password__form"
      onSubmit={[Function]}
    >
      <div
        className="input__container"
      >
        <h4
          className="input__text input__text--invisible"
        >
          Your new password must be different from the old password.
        </h4>
        <input
          className="input__field input__field--margin-top"
          onChange={[Function]}
          placeholder="Old Password"
          type="password"
          value=""
        />
      </div>
      <div
        className="input__container"
      >
        <h4
          className="input__text input__text--invisible"
        >
          Password must has 1 lowercase, 1 uppercase letter, 1 number and at least 8 characters in length
        </h4>
        <input
          className="input__field input__field--margin-top"
          onChange={[Function]}
          placeholder="New Password"
          type="password"
          value=""
        />
      </div>
      <div
        className="input__container"
      >
        <h4
          className="input__text input__text--invisible"
        >
          Password not match
        </h4>
        <input
          className="input__field input__field--margin-top"
          onChange={[Function]}
          placeholder="Confirm new password"
          type="password"
          value=""
        />
      </div>
      <button
        className="primary-button "
      >
        <span>
          Submit
        </span>
        <div
          className="spinner-border spinner-border-m text-light"
        />
      </button>
    </form>
  </div>
</div>
`;
